### PR TITLE
Replace CONTENT_TYPE String with Vertx optimized one and make fast-path inlineable

### PR DESCRIPTION
**Pull request by franz1981**
The original method was not inlineable due to its bytecode size being > `FreqInlineSize` (which is 325 bytes).
Splitting it into fast/slow path make it inlineable.
Furthermore I'm replacing the `String` upper-case header name to allow vertx headers to reuse its precomputed hashcode.

@geoand is it safe to assume that it's always a Vertx HTTP response?

I see a comment stating that we have to use it upper-case, is it still valid?
Considering that the map is case-insensitive, I don't understand why...

### CODE REVIEW

**Comment by geoand** [comment]
The core module does not depen on Vert.x, so we'll have to find another to integrate this improvement

**Comment by franz1981** [comment]
Got it so, having on https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/spi/ServerHttpResponse.java#L10 a specific method for Content type eg `setContentType(CharSequence value)` can help the implementors to have an optimized version for it, but I need first to understand this

> I see a comment stating that we have to use it upper-case, is it still valid?
Considering that the map is case-insensitive, I don't understand why...

In any case, the additional method (and `getAccept(boolean single)` could be anther one, given thst it pop up frequently in profiling data) sadly will look redundant, having already a generic method for it, but the same applies to Vertex Multimap API which have a mix of String/CharSequence types floating around (due to the original bad design decisions of Netty header map), making this "optimized" API methods not a novelty.

A different option, but more "disruptive" could be to define an HeaderName interface type (which have an associated monotonic integer id) with an implementation using an enum which list all possible header names which expose an id based on the enum ordinal: this can be used to create an array of AsciiString to perform O(1) translation between this API ans vertx one.
This new interface can both be used to:
- pair with the existing one using strings for "known header names" 
- replace the existing one, but specifying an additional concrete type to "extend" the enum with custom values (which id won't exist, or return a negative value)

In case for this current use case and using the latter option, if we really have to specify a uppercase header name,given that we want to benefit from caching the right AsciiString, we need to make it part of the enum (which is pretty ugly). I still have to think how to cover this use case properly, but if we don't care for it, would be simpler.

**Comment by geoand** [comment]
> I see a comment stating that we have to use it upper-case, is it still valid?

Probably, but we can certainly make a change and run the tests to make sure.

File: independent-projects/resteasy-reactive/server/runtime/pom.xml [hunk 1]
```diff
@@ -19,6 +19,10 @@
             <groupId>io.quarkus.resteasy.reactive</groupId>
             <artifactId>resteasy-reactive-common</artifactId>
         </dependency>
+        <dependency>
```
Comment (cescoffier) [inline independent-projects/resteasy-reactive/server/runtime/pom.xml:22]: Do we really want this? I'm not against it, obviously, but it's a big change. 

@geoand

Comment (geoand) [reply independent-projects/resteasy-reactive/server/runtime/pom.xml:22]: We do **not** want it. 

I to Franz that the core module does not depend on Vert.x and I don't want to add such a dependency just for the sake of this change

**Comment by franz1981** [comment]
I'm replacing few commits here for a different proposal, to address a wider performance issue

**Comment by geoand** [comment]
Can we please squash the third commit?

**Review by geoand — APPROVED** [review]

**Comment by franz1981** [comment]
@geoand I'm currently sending few HTTP 2 improvements in Netty and Vertx and this one popup again, but in a different form eg

image


In short, I see 2 problems, related header names:
1. vertx play safe (it's an hard requirement), and transform http headers in lower-case form, while JAX-RS it is not eg "Content-Type"
2. the vertx lower-case form preserve the header name type (eg `String` is still `String`, `AsciiString` is `AsciiString`), see https://github.com/eclipse-vertx/vert.x/pull/5055
3. Netty requires 2 different code paths to perform ASCII hash-code computation: cached `AsciiString` have a cached `hash` which is valid to index into Netty data structures (mostly `DefualtHttp2Headers`), but it forces to compute a new one of the fly, each time is a `String`, not benefitting of caching. 

Usually this is not a biggie, if not that introduce both allocations and computations which are not really needed.

The second problem instead is on the header value (and this is beyond the scope of this PR, still under evaluation), HTTP 2 introduces HPack static tables to skip re-sending known header/name values over the wire 
eg `content-type`:`text/plain;charset=UTF-8` and Netty itself, try hard to reuse the known ones, by caching them

The original HTTP 2 caching (within Netty) was just on header names, at https://github.com/netty/netty/blob/aa07be47d9d2bbb09b1ca9897575ebc4de6056c0/codec-http2/src/main/java/io/netty/handler/codec/http2/HpackEncoder.java#L105 with a default of 4096 entries, and has been improved recently by https://github.com/netty/netty/pull/12620, adding name-value pairs, which requires computing ascii hash code (again) on values as well (!!!): this translate again in computing ascii hash code in the hot path for contant 'String`s like `text/plain;charset=UTF-8` (which we store in https://github.com/quarkusio/quarkus/blob/004d51df9f55c4c646ac8a7fe98e0ab32e20a27f/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ServerSerialisers.java#L488, hence related to the `MediaType`).

In short, the fact that Netty/Vertx prefer `AsciiString`s for both header name/values (possibly constant, to save computing the same ascii hash code over and over and saving lower-casing them) become even more important with Http 2 and I really think we should address this, somehow.

**Comment by geoand** [comment]
I'm definitely open to ideas for improvement :)

**Comment by geoand** [comment]
> it really is mandatory?

I don't remember, but my guess is that some test will fail if changed. That said, it's definitely worth running CI with the best implementation to see what (if anything) fails

**Comment by franz1981** [comment]
Yeah, consider that I found this, while testing the benefits of this PR: https://github.com/netty/netty/pull/13824

in short, if we now use "known" Netty header names we would still get some equality check in the hot path because `accept` isn't `Accept` and Netty requires to perform case-indipendent equality checks char by char , see 

image

which violet part is `org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.getAccept`

**Comment by geoand** [comment]
Unfortunately we do have to live with some of the messy world limitations...

**Comment by gsmet** [comment]
Where are we on this one? If we don't expect to be able to make it work, we should close it. Thanks!

**Comment by cescoffier** [comment]
Agreed with @gsmet.

Do we have a path forward?

**Comment by gastaldi** [comment]
Closing it for now, let's reopen once it's rebased and has a path forward.
